### PR TITLE
GKE Update

### PR DESCRIPTION
**Pull request by ichekrygin**
While GKE is in need of a more comprehensive update, this PR attempts to alleviate specific pain-points:

- Apply "new design paradigm" `reconcileAfter`                                                       
- Fix Location column name                                                                           
- Add Resource Labels support (client/controller)           

Resolves #

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make generate`) has been run to update object specifications, if necessary.
- [ ] CRD manifests generation (`make manifests`) has been run to update CRD manifests yaml file specifications, if necessary.
- [ ] Update dependencies (`make vendor`) has been run to update `Gopkg.lock` file, if necessary
- [ ] RBAC permissions in `clusterrole.yaml` have been updated to include new types, if necessary
- [ ] All related commits have been squashed to improve readability.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

File: pkg/controller/gcp/compute/gke.go [hunk 1]
```diff
@@ -21,9 +21,11 @@ import (
 	"encoding/base64"
 	"fmt"
 
+	"github.com/pkg/errors"
+
```
Comment (negz) [inline pkg/controller/gcp/compute/gke.go]: Looks like some whitespace snuck in here.

File: pkg/controller/core/resource_controller.go [hunk 2]
```diff
@@ -46,6 +47,11 @@ const (
 	errorSettingResourceBindStatus   = "Failed to set resource binding status"
 	errorResettingResourceBindStatus = "Failed to reset resource binding status"
 	waitResourceIsNotAvailable       = "Waiting for resource to become available"
+
+	// RequeueOnWait - requeue after duration when waiting for resource state
+	RequeueOnWait = 30 * time.Second
+	// RequeueOnSuccess - requeue after duration when resources status is successful
+	RequeueOnSuccess = 2 * time.Minute
```
Comment (negz) [inline pkg/controller/core/resource_controller.go]: Nice. I'll update my controllers to use these if this gets merged today.

File: pkg/controller/gcp/compute/gke.go [hunk 3]
```diff
@@ -191,12 +201,11 @@ func (r *Reconciler) _create(instance *gcpcomputev1alpha1.GKECluster, client gke
 	}
 
 	instance.Status.State = gcpcomputev1alpha1.ClusterStateProvisioning
-
 	instance.Status.UnsetAllConditions()
 	instance.Status.SetCreating()
 	instance.Status.ClusterName = clusterName
 
-	return resultRequeue, r.Update(ctx, instance)
+	return reconcile.Result{}, errors.Wrapf(r.Update(ctx, instance), updateErrorMessageFormat, instance.GetName())
```
Comment (negz) [inline pkg/controller/gcp/compute/gke.go]: Do we really not want to requeue here, or should this be a RequeueOnSuccess scenario?

File: pkg/controller/gcp/compute/gke.go [hunk 4]
```diff
@@ -251,12 +260,9 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
...
 	err := r.Get(ctx, request.NamespacedName, instance)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+		if kerrors.IsNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
```
Comment (negz) [inline pkg/controller/gcp/compute/gke.go]: Do we really not want to requeue here, or should this be a RequeueOnFailure scenario?

File: pkg/controller/gcp/compute/gke.go (hunk 3 above)
Comment (ichekrygin) [reply pkg/controller/gcp/compute/gke.go]: In this context the `requeue` is superfluous, the object will be requeued due to the update ("dirty state"). If update operations itself fails - the object will be requeued due to the error.

File: pkg/controller/gcp/compute/gke.go (hunk 4 above)
Comment (ichekrygin) [reply pkg/controller/gcp/compute/gke.go]: This is similar to the case above: the explicit `requeue` in this case is not needed since the object will be requeued due to the error.

**Comment by ichekrygin** [comment]
Note: as it stands this PR has low code coverage: https://sonarcloud.io/component_measures?id=crossplaneio_crossplane&metric=new_coverage&pullRequest=414&view=list

image

This attributed to two factors/files:
1. `pkg/clients/gcp/gke/gke.go` has 0% coverage on the new code - this is the integration piece with the Google API "google.golang.org/api/container/v1". I think there could be a valid case to exclude this specific file, due to very little functional code, and instead mostly a "glue" with the google library.

1. `pkg/controller/gcp/compute/gke.go` - has low `62.5%` coverage, and this is valid case where coverage should(must) be increased. However, taking to consideration that this controller itself needs to be refactored (to adhere to new standards), in addition, that the tests need to be re-written to match the controller change as well as to be "Table Driven" as our newer examples, I dare to propose that we can approve this with PR with insufficient coverage in lew of addressing this issue in subsequent PR(s).

File: pkg/controller/gcp/compute/gke.go (hunk 3 above)
Comment (negz) [reply pkg/controller/gcp/compute/gke.go]: I had not considered that calling `Update` on the same resource we're reconciling would trigger another reconcile, regardless of the `reconcile.Result{}` we return, but that makes sense.

If I understand correctly there are a few different cases we may have to handle. I wonder if we can add some constants to the `core` package that help make these cases really obvious. Something like:

```go
// Reconcile results.
const (
  // ResultResourceMutating should be returned when the Reconciler is waiting for an
  // operation to complete, for example waiting for the managed resource the  Kubernetes
  // resource represents to update. It will requeue a Reconcile after a short duration.
  ResultResourceMutating = reconcile.Result{RequeueAfter: 30 * time.Second}

  // ResultResourceClean should be returned when the Reconciler is confident it did
  // not alter the Kubernetes resource, for example because it is in sync with the
  // managed resource it represents. It will requeue a Reconcile after a long duration.
  ResultResourceClean = reconcile.Result{RequeueAfter: 2 * time.Minute}

  // ResultResourceDirty should be returned when the Reconciler is confident it has
  // altered the Kubernetes resource. In this case a Reconcile will be queued implicitly
  // when the dirty resource is updated, so there's no need for us to trigger one.
  ResultResourceDirty = reconcile.Result{}

  // ResultResourceMaybeDirty should be returned when the Reconciler might have altered
  // the Kubernetes resource. We'll trigger a Reconcile after a long duration whether the
  // resource is clean or dirty. If the resource is dirty we'll also trigger an immediate
  // reconcile. Be wary of creating an infinite reconcile loop when using this result.
  ResultResourceMaybeDirty = ResultResourceClean

  // ResultTransientError should be returned when the Reconciler encounters what it
  // expects to be a transient error from the API server.
  ResultTransientError = reconcile.Result{RequeueAfter: 5 * time.Second}

  // ResultPermanentError should be returned when the Reconciler encounters what it
  // expects to be a permanent error - one that is likely to recur consistently until fixed.
  ResultPermanentError = ResultResourceClean
)
```

What do you think?

Comment (ichekrygin) [reply pkg/controller/gcp/compute/gke.go]: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L210

As you can see, reconciliation requeue is a little more subtle. I tried to capture this in my "Reconcile Patterns" document. Here is copy & paste from it:

Object State | Reconcile. Result | Error | When to Requeue
-- | -- | -- | --
Dirty | Any | Any | Immediate.
Clean | Any | Error | Exponential backoff
Clean | Result{Requeue: true} | Any | Exponential backoff
Clean | Result{Requeue: false} | Nil | Sync Period
Clean | Result{RequeueAfter: 1 minute} | Nil | After 1 minute delay

To elaborate. Dirty state and Error have equal precedence, with Dirty state being on a little higher end by virtue (excluding external update), the controller must successfully update the object. If the object was updated (externally or by the controller), it is immediately requeued irrespective of error. To emphasize this point it `reconcile.Result{Requeue or RequeueAfter}` is entirely ignored if the object is either dirty or reconciliation returned an error.  The key difference between "dirty" and "error" is in the latter case the object is an object is queued via `RateLimited` interface, i.e. exponential-backoff.

`reconcile.Result` have two options: `Requeue: bool` and `RequeueAfter: time.Duration`. The latter takes precedence and uses `DelayingInterface`, wheres the former uses the same `RateLimited` interface as in the case with returning the error, `reconcile.Result{Requeue: true}` is identical in terms of handling as if the Reconcile function had returned an error. 

With that out of the way, I think we don't need:
```go
  // ResultTransientError should be returned when the Reconciler encounters what it
  // expects to be a transient error from the API server.
  ResultTransientError = reconcile.Result{RequeueAfter: 5 * time.Second}

  // ResultPermanentError should be returned when the Reconciler encounters what it
  // expects to be a permanent error - one that is likely to recur consistently until fixed.
  ResultPermanentError = ResultResourceClean
```
An instead I think we do want either return the error or `reconcile.Result{Requeue: true}`

To summarize: 

1. The only time we want to return:
```go
reconcile.Result{}, nil
```
is when we are Done with this object and the only case for this object to be processed again is either on global sync or external update.

1. If we are returning the error, then it doesn't matter what `reconciler.Result` is, i.e. 
```go
reconcile.Result{}, err
```
is semantically equivalent to:
```go
reconcile.Result{Requeue: true}, err
```
or
```go
reconcile.Result{RequeueAfter: time}, err
```

1. Typically the only time we want to return:
```go
reconcile.Result{Requeue: true}, nil
```
is when we handled the processing error (i.e. errors other than Update or Status().Update) and activated `Failed` condition

Comment (negz) [reply pkg/controller/gcp/compute/gke.go]: Aha, I forgot that you worked through all these cases already.

All the different cases aside, my main point is that I think some well named constants representing different 'meanings' of `reconcile.Result{...}` variants would help explain the intent of the code a lot. In this specific case returning `core.ResultDirty` (which would have GoDoc explaining what pointed out in this thread) would help the code document itself.
